native-activity/app/src/main/cpp/main.cpp

```diff
@@ -35,9 +36,34 @@
 #define _LOG(priority, fmt, ...) \
   ((void)__android_log_print((priority), (LOG_TAG), (fmt)__VA_OPT__(, ) __VA_ARGS__))
 
+#define LOGE(fmt, ...) _LOG(ANDROID_LOG_ERROR, (fmt)__VA_OPT__(, ) __VA_ARGS__)
 #define LOGW(fmt, ...) _LOG(ANDROID_LOG_WARN, (fmt)__VA_OPT__(, ) __VA_ARGS__)
 #define LOGI(fmt, ...) _LOG(ANDROID_LOG_INFO, (fmt)__VA_OPT__(, ) __VA_ARGS__)
 
+[[noreturn]] __attribute__((__format__(__printf__, 1, 2))) static void fatal(
+    const char* fmt, ...) {
+  va_list ap;
+  va_start(ap, fmt);
+  char buf[BUFSIZ];
+  if (vsnprintf(buf, sizeof(buf), fmt, ap) < 0) {
+    android_set_abort_message("failed for format error message");
```

Comment:
vasprintf()? you're not assuming heap corruption here, right?

Reply:
No, I just always forget about that variant.